### ISSUE TITLE
BOSH connection closed after a couple of seconds

### ISSUE THREAD

**Issue by nunohorta** [processone/ejabberd#835]
Hi, I'm trying to use bosh with ejabberd. Here's a part of the configuration:

hosts:
    - "localhost"
    - "example.domain.com"
listen:
    - { port: 5222, module: ejabberd_c2s, max_stanza_size: 65536, shaper: c2s_shaper, access: c2s }
    - { port: 5269, module: ejabberd_s2s_in }
    - { port: 5280, module: ejabberd_http, request_handlers: {  }, web_admin: true, http_bind: true, captcha: true }
    - { port: 4560, module: ejabberd_xmlrpc }
......
and the module
...
modules:
  mod_http_bind: {}

Connection and authentication is working fine but after a while(seconds) the connection is closed. 
The logs can be found here: https://gist.github.com/nunohorta/cf05baee3000dcfeb3e9
There's a crash report but I'm not sure what it means.

**Comment by prefiks** [comment]
So the problematic fragment is this:

```
2015-11-12 12:37:01.829 [error] <0.552.0> CRASH REPORT Process <0.552.0> with 0 neighbours crashed with reason: no function clause matching prim_inet:recv0(#Port<0.18834>, -855, 300000)
```

and it fails because it doesn't like negative number as second parameter.

I'm guessing this is called from `ejabberd_http:recv_data()` but have no idea why this calls with negative value, is client sending that in `Content-Length` header?

Could you describe client/environment you are using to test this? Could you please also check error.log file in ejabberd log place to see if there is more info about this?

**Comment by nunohorta** [comment]
I'm using pidgin to connect to the server. The error.log shows the same error:

`2015-11-12 13:39:55.690 [error] <0.608.0> CRASH REPORT Process <0.608.0> with 0 neighbours crashed with reason: no function clause matching prim_inet:recv0(#Port<0.18965>, -471, 300000)`
`2015-11-12 13:39:55.690 [error] <0.298.0> Supervisor ejabberd_http_sup had child undefined started with {ejabberd_http,start_link,undefined} at <0.608.0> exit with reason function_clause in context child_terminated`

I have a tcpdump file captured when this occurs. I can send it to you if it helps.

**Comment by prefiks** [comment]
Sure that would be great, you can also make bump loglevel in ejabberd (set to `loglevel: 5` in ejabberd.yml) to get more output from it.

**Comment by nunohorta** [comment]
thanks! Here's the tcpdump file:
https://www.dropbox.com/s/wi61mt1ifebtg0t/dump.pcap?dl=0
and here is ejabberd logs with loglevel set to 5:
https://www.dropbox.com/s/fwd84brtcz95m36/ejabberd.log?dl=0

**Comment by prefiks** [comment]
Thank you, i will analyze those, and update this ticket when i found something.

**Comment by nunohorta** [comment]
thank you!

**Comment by prefiks** [comment]
Fix for this issue pushed as commit 3ee5195b7a7dc1ac19396a8f3e289b468f635071. If you are building your own ejabberd from source, updating from git and compiling ejabberd_http.erl should help with this issue, if you are using installer, unfortunately you would need to wait for next release.

**Comment by nunohorta** [comment]
nice, thank you very much!
